PSR7 IntegrationTestCase part 1

markstory opened the pull request:
This is the start of adding PSR7 compatibility to `IntegrationTestCase`. I feel being able to do full-stack integration testing is pretty important for PSR7 to be a really viable option for developers.
### Breaking Changes

This contains a breaking change for the `_buildRequest` method. I feel this is a reasonable change to make as it is a protected method. What I needed was a way to get the request parameters not in an object form. Trying to extract all the headers and other data out of a `Cake\Network\Request` was not looking feasible because of how headers are implemented.
### Questions
- Is the general approach of using 'strategy' adapters for each HTTP stack a reasonable approach?
- Is the testing approach looking reasonable, or should I create another test case file with all the PSR7 specific tests?
- Does auto-detecting the `Application` class seem like a reasonable way to behave? Or should using PSR7 always require a call to `useHttpServer()`?
- Do the classnames I've chosen make sense? Are there better names that you can suggest?]
- How should sessions be handled? I was considering using a PSR7 attribute.

Refs #6960

markstory committed:
Start sketching out the PSR7 IntegrationTest work.

The biggest hurdle is how to invoke either the existing Dispatcher stack
or the Http\Server stack. I'm going to try sniffing out the class name,
and use its presence to go into PSR7 mode.

The PSR7 mode will do an additional translation from cake->psr as
I don't want to try and re-jig all the internals of IntegrationTestCase
just yet.

markstory committed:
Rough out Middleware Dispatcher

This is a very very rough prototype of how PSR7 request dispatching could be
integrated into into the TestSuite. There is still a lot TODO, but
a very basic GET request is somewhat functional right now.

markstory committed:
Make IntegrationTestCase::_buildRequest return an array.

This is a breaking change to the protected method, but I feel its
necessary in order to make other things easier. For example, getting all
headers out of a Cake\Request is not possible right now. By using the
array return value we can easily access all the different aspects of the
simulated request.

dereuromark committed:
Small fixes.

markstory committed:
Remove unused import.

chinpei215 commented on src/TestSuite/MiddlewareDispatcher.php:
```diff
@@ -0,0 +1,82 @@
+
+        $server = new Server($app);
+
+        // TODO How to pass session data? PSR7 requests don't handle sessions..
```
If so, how do we use `RequestTransformer::toCake()` with session?

markstory replied:
I'm thinking we might be able to put the session in as a request 'attribute' like `$request->withAttribute('session', $session)`, and then pull that out in the transformer.

lorenzo commented on src/TestSuite/IntegrationTestCase.php:
```diff
@@ -352,17 +384,15 @@ public function delete($url)
-            [$this, 'controllerSpy']
-        );
+        if ($this->_useHttpServer) {
+            $dispatcher = new MiddlewareDispatcher($this);
```
Maybe get this from a protected method? So that we can offer the ability to override this object.

markstory replied:
Good idea.

lorenzo commented on src/TestSuite/MiddlewareDispatcher.php:
```diff
@@ -0,0 +1,82 @@
+        try {
+            $namespace = Configure::read('App.namespace');
+            $reflect = new ReflectionClass($namespace . '\Application');
+            $app = $reflect->newInstance('./config');
```
is is always going to be `./config` ? Or do people have a way to change that?

markstory replied:
People do have a way to change it, but I don't know how to propagate that here.

lorenzo commented:
> Is the general approach of using 'strategy' adapters for each HTTP stack a reasonable approach?

The implementation looks very simple and elegant, I think it is a good idea.

> Is the testing approach looking reasonable, or should I create another test case file with all the PSR7 specific tests?

It's reasonable, I would not complicate it more.

> Does auto-detecting the Application class seem like a reasonable way to behave? Or should using PSR7 always require a call to useHttpServer()?

I'm more inclined to say that it should always require it, but given how easy it is to detect and support, it seems like a non-issue.

> Do the classnames I've chosen make sense? Are there better names that you can suggest?

Given that the RequestDispatcher is for legacy maybe call it `LegacyRequestDispatcher`? 

> How should sessions be handled? I was considering using a PSR7 attribute.

I always though them as an attribute of the request, what are other frameworks doing?

markstory commented:
> I always though them as an attribute of the request, what are other frameworks doing?

It seems that Slim3 doesn't include sessions, at least their docs don't cover how sessions work, and I wasn't able to find any code for sessions either. There are [flash message services](https://github.com/slimphp/Slim-Flash) that use sessions, but they inject the session onto the route handlers.

Zend-expressive uses [session middleware](https://github.com/dannym87/expressive-session-middleware) to attach an attribute onto the request as I've suggested here.

markstory committed:
Update PSR7 integration tests based on feedback.

markstory commented:
I'm planning on implementing sessions and input stream support in the next few days, but I'll likely do a follow up pull request with more test cases as I'm not comfortable with the present coverage and don't want to make a gigantic pull request.

lorenzo commented:
@markstory Should be remove the `TODO` rule in the style checker?

markstory commented:
@lorenzo I will get the TODO's removed tonight. In general I think TODO comments are a smell we should avoid.

markstory committed:
Allow PSR7 requests to store the session as an attribute.

By allowing the session to exist on the PSR7 request, it will be easier
to implement integration test case flows and more importantly let people
access the session from their psr7 requests.

markstory committed:
Add session support for PSR7 requests.

markstory committed:
Remove TODOs

The base and webroot attributes are handled by the RequestTransformer.
I'm not convinced that php://input needs to be handled as the legacy
dispatcher doesn't do a great job of it either.

markstory committed:
Allow request bodies to be transformed.

This is useful for API based applications.

markstory committed:
Add integration tests for PSR7 + request bodies.

markstory commented:
Now with sessions and request body payloads.

lorenzo commented on src/Http/RequestTransformer.php:
```diff
@@ -48,6 +48,12 @@ public static function toCake(PsrRequest $request)
 
+        $input = null;
+        $stream = $request->getBody();
+        if ($stream->getSize()) {
```
After knowing a bit of how a stream is implemented, I'd suggest just getting the content and then casting the `$input` to null if it is an empty string

markstory replied:
I don't completely understand. Cake\Network\Request expects a string for the 'input' if its being injected, which is what this is trying to do.

lorenzo replied:
This is what I meant:
```php
$input = $stream->geContents();
$input = $input === '' ? null : $input;
```
 My goal is to save the call to `getSize()`, since in many stream implementations that is done by literally counting the bytes one by one in php

markstory replied:
Ah ok. I can fix that up.

lorenzo commented:
Great job mark!

markstory committed:
Don't call getSize()

getSize() can be expensive, instead read the stream and conditionally
assign the input with the stream contents.
